File: LeetCode/Hao/92ReverseLinkedListII.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if (head == null || head.next == null) {
            return head;
        }
        if (n == m) {
            return head;
        }
        
        ListNode p = new ListNode(1);
        p.next = head;
        ListNode ans = p;
        
        for (int i = 1; i < m; i ++) {
            p = p.next;
        }
        
        ListNode pre = p.next;
        ListNode cur = p.next.next;
        
        for (int i = 0; i < n - m; i ++) {
            ListNode tmp = cur.next;
            cur.next = pre;
            pre = cur;
            cur = tmp;
        }
        
        ListNode tmp = p.next;
        p.next = pre;
        tmp.next = cur;
        
        return ans.next;
    }
}

File: LeetCode/Peng/92ReverseLinkedListII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

//Use dummy head will be easy to handle boundries.
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (!head || !head->next || m == n) return head;
        
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* last = dummy;
        ListNode* current, * next, * left, * right;
        for (int i = m - 1; i > 0 ; i--) {
            last = last->next;
        }
        left = last;
        right = last->next;
        current = last->next;
        next = current->next;
        
        for (int i = n - m; i > 0; i--) {
            last = current;
            current = next;
            next = next->next;
            current->next = last;
        }
        right->next = next;
        left->next = current;
        return dummy->next;
    }
};